### PR TITLE
Refactor and Tests (in prgress)

### PR DESCRIPTION
**Pull request by minhptruong**

### CODE REVIEW

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
- Added 162
           

Complexity increasing per file
==============================
- packages/arb-validator-core/ethbridgetest/protocoltester/protocoltester.go  3
- packages/arb-validator-core/ethbridgetest/protocol_test.go  3
- packages/arb-validator/challenges/challengeUtils.go  3
- packages/arb-validator-core/ethbridgetest/value_test.go  6
- packages/arb-validator/challenges/messagesChallenger_test.go  4
- packages/arb-validator/challenges/messagesChallenger.go  10
- packages/arb-validator/proofmachine/machinetester/machinetester.go  3
- packages/arb-validator/challenges/executionDefender.go  10
- packages/arb-validator/proofmachine/machine_test.go  5
- packages/arb-validator-core/ethbridgetest/main_test.go  4
- packages/arb-validator/challenges/messagesDefender.go  12
- packages/arb-validator/challenges/inboxTopChallenger.go  9
- packages/arb-validator/challenges/inboxTopDefender.go  9
- packages/arb-validator/challenges/executionChallenger.go  10
         

Complexity decreasing per file
==============================
+ packages/arb-validator/challenges/inboxTopChallenge_test.go  -1
+ packages/arb-validator/challenges/executionChallenge_test.go  -1
+ packages/arb-validator/challenges/testHelper.go  -10
+ packages/arb-validator-core/ethbridge/arbRollup.go  -1
         

Clones added
============
- packages/arb-validator-core/ethbridgetest/protocoltester/protocoltester.go  16
- packages/arb-validator/challenges/challengeUtils.go  14
- packages/arb-validator-core/ethbridgetest/value_test.go  2
- packages/arb-validator-core/ethbridgetest/messagetester/messagetester.go  3
- packages/arb-validator-core/ethbridgetest/valuetester/valuetester.go  5
- packages/arb-validator/challenges/messagesChallenger.go  6
- packages/arb-validator/proofmachine/machinetester/machinetester.go  13
- packages/arb-validator/challenges/executionDefender.go  4
- packages/arb-validator/challenges/messagesDefender.go  6
- packages/arb-validator/challenges/inboxTopChallenger.go  3
- packages/arb-validator/challenges/inboxTopDefender.go  2
- packages/arb-validator/challenges/executionChallenger.go  4
         

Clones removed
==============
+ packages/arb-validator/challenges/inboxTopChallenge_test.go  -11
+ packages/arb-validator/challenges/testHelper.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5708250&bid=18682118)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 18
- Added 179
           

Complexity increasing per file
==============================
- packages/arb-validator/challenges/messagesChallenger_test.go  4
- packages/arb-validator-core/ethbridgetest/value_test.go  6
- packages/arb-util/value/hashPreImage.go  1
- packages/arb-validator-core/ethbridgetest/main_test.go  4
- packages/arb-validator/challenges/executionDefender.go  10
- packages/arb-validator/challenges/executionChallenger.go  10
- packages/arb-validator/challenges/messagesChallenger.go  10
- packages/arb-validator-core/ethbridgetest/protocoltester/protocoltester.go  3
- packages/arb-validator/challenges/messagesDefender.go  12
- packages/arb-validator-core/ethbridgetest/protocol_test.go  3
- packages/arb-validator/proofmachine/machine_test.go  5
- packages/arb-validator/challenges/inboxTopChallenger.go  9
- packages/arb-avm-go/checkpoint/checkpointValues.go  1
- packages/arb-validator/proofmachine/machinetester/machinetester.go  3
- packages/arb-validator/challenges/inboxTopDefender.go  9
- packages/arb-validator/challenges/challengeUtils.go  3
         

Complexity decreasing per file
==============================
+ packages/arb-avm-go/vm/machinestate.go  -7
+ packages/arb-validator/challenges/testHelper.go  -10
+ packages/arb-validator-core/ethbridge/arbRollup.go  -1
+ packages/arb-validator/challenges/inboxTopChallenge_test.go  -1
+ packages/arb-validator/challenges/executionChallenge_test.go  -1
         

Clones added
============
- packages/arb-validator-core/ethbridgetest/value_test.go  2
- packages/arb-validator-core/ethbridgetest/messagetester/messagetester.go  3
- packages/arb-validator/challenges/executionDefender.go  4
- packages/arb-validator/challenges/executionChallenger.go  4
- packages/arb-validator/challenges/messagesChallenger.go  6
- packages/arb-validator-core/ethbridgetest/protocoltester/protocoltester.go  16
- packages/arb-validator-core/ethbridge/messagesChallenge.go  2
- packages/arb-validator/challenges/messagesDefender.go  6
- packages/arb-validator/challenges/inboxTopChallenger.go  3
- packages/arb-validator-core/ethbridgetest/valuetester/valuetester.go  5
- packages/arb-validator/proofmachine/machinetester/machinetester.go  13
- packages/arb-validator/challenges/inboxTopDefender.go  2
- packages/arb-validator/challenges/challengeUtils.go  14
         

Clones removed
==============
+ packages/arb-validator/challenges/testHelper.go  -1
+ packages/arb-validator/challenges/inboxTopChallenge_test.go  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5708250&bid=18682118)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 11
- Added 179
           

Complexity increasing per file
==============================
- packages/arb-validator/challenges/messagesDefender.go  12
- packages/arb-validator/challenges/messagesChallenger.go  10
- packages/arb-validator-core/ethbridgetest/protocol_test.go  3
- packages/arb-validator/challenges/inboxTopChallenger.go  9
- packages/arb-validator/challenges/inboxTopDefender.go  9
- packages/arb-validator-core/ethbridgetest/protocoltester/protocoltester.go  3
- packages/arb-validator/proofmachine/machinetester/machinetester.go  3
- packages/arb-validator/challenges/executionDefender.go  10
- packages/arb-validator/proofmachine/machine_test.go  5
- packages/arb-util/value/hashPreImage.go  3
- packages/arb-validator/challenges/messagesChallenger_test.go  4
- packages/arb-validator-core/ethbridgetest/main_test.go  4
- packages/arb-validator-core/ethbridgetest/value_test.go  6
- packages/arb-validator/challenges/executionChallenger.go  10
- packages/arb-validator/challenges/challengeUtils.go  3
         

Complexity decreasing per file
==============================
+ packages/arb-validator/loader/loader.go  -1
+ packages/arb-validator/challenges/testHelper.go  -10
+ packages/arb-validator/challenges/inboxTopChallenge_test.go  -1
+ packages/arb-validator-core/ethbridge/arbRollup.go  -1
+ packages/arb-validator/challenges/executionChallenge_test.go  -1
         

Clones added
============
- packages/arb-validator-core/ethbridge/messagesChallenge.go  2
- packages/arb-validator/challenges/messagesDefender.go  6
- packages/arb-validator-core/ethbridgetest/valuetester/valuetester.go  5
- packages/arb-validator/challenges/messagesChallenger.go  6
- packages/arb-validator/proofmachine/proofmachine.go  1
- packages/arb-validator/challenges/inboxTopChallenger.go  3
- packages/arb-validator/challenges/inboxTopDefender.go  2
- packages/arb-validator-core/ethbridgetest/protocoltester/protocoltester.go  16
- packages/arb-validator/proofmachine/machinetester/machinetester.go  13
- packages/arb-validator/challenges/executionDefender.go  4
- packages/arb-validator-core/ethbridgetest/value_test.go  2
- packages/arb-validator/challenges/executionChallenger.go  4
- packages/arb-validator-core/ethbridgetest/messagetester/messagetester.go  3
- packages/arb-validator/challenges/challengeUtils.go  14
         

Clones removed
==============
+ packages/arb-validator/challenges/testHelper.go  -1
+ packages/arb-validator/challenges/inboxTopChallenge_test.go  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5708250&bid=18682118)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 11
- Added 179
           

Complexity increasing per file
==============================
- packages/arb-validator/challenges/inboxTopChallenger.go  9
- packages/arb-validator-core/ethbridgetest/main_test.go  4
- packages/arb-util/value/hashPreImage.go  3
- packages/arb-validator-core/ethbridgetest/protocol_test.go  3
- packages/arb-validator/challenges/executionChallenger.go  10
- packages/arb-validator/challenges/messagesDefender.go  12
- packages/arb-validator/challenges/executionDefender.go  10
- packages/arb-validator-core/ethbridgetest/value_test.go  6
- packages/arb-validator/challenges/messagesChallenger_test.go  4
- packages/arb-validator/proofmachine/machine_test.go  5
- packages/arb-validator-core/ethbridgetest/protocoltester/protocoltester.go  3
- packages/arb-validator/challenges/challengeUtils.go  3
- packages/arb-validator/challenges/inboxTopDefender.go  9
- packages/arb-validator/proofmachine/machinetester/machinetester.go  3
- packages/arb-validator/challenges/messagesChallenger.go  10
         

Complexity decreasing per file
==============================
+ packages/arb-validator/challenges/executionChallenge_test.go  -1
+ packages/arb-validator/loader/loader.go  -1
+ packages/arb-validator/challenges/inboxTopChallenge_test.go  -1
+ packages/arb-validator/challenges/testHelper.go  -10
+ packages/arb-validator-core/ethbridge/arbRollup.go  -1
         

Clones added
============
- packages/arb-validator/proofmachine/proofmachine.go  1
- packages/arb-validator/challenges/inboxTopChallenger.go  3
- packages/arb-validator/challenges/executionChallenger.go  4
- packages/arb-validator/challenges/messagesDefender.go  6
- packages/arb-validator-core/ethbridge/messagesChallenge.go  2
- packages/arb-validator/challenges/executionDefender.go  4
- packages/arb-validator-core/ethbridgetest/valuetester/valuetester.go  5
- packages/arb-validator-core/ethbridgetest/value_test.go  2
- packages/arb-validator-core/ethbridgetest/protocoltester/protocoltester.go  15
- packages/arb-validator-core/ethbridgetest/messagetester/messagetester.go  2
- packages/arb-validator/challenges/challengeUtils.go  14
- packages/arb-validator/challenges/inboxTopDefender.go  2
- packages/arb-validator/proofmachine/machinetester/machinetester.go  13
- packages/arb-validator/challenges/messagesChallenger.go  6
         

Clones removed
==============
+ packages/arb-validator/challenges/inboxTopChallenge_test.go  -11
+ packages/arb-validator/challenges/testHelper.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5708250&bid=18682118)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 11
- Added 179
           

Complexity increasing per file
==============================
- packages/arb-validator/challenges/inboxTopChallenger.go  9
- packages/arb-validator/challenges/messagesDefender.go  12
- packages/arb-validator-core/ethbridgetest/protocoltester/protocoltester.go  3
- packages/arb-validator/challenges/inboxTopDefender.go  9
- packages/arb-validator/challenges/messagesChallenger_test.go  4
- packages/arb-validator-core/ethbridgetest/protocol_test.go  3
- packages/arb-validator/challenges/messagesChallenger.go  10
- packages/arb-validator/proofmachine/machine_test.go  5
- packages/arb-validator-core/ethbridgetest/main_test.go  4
- packages/arb-util/value/hashPreImage.go  3
- packages/arb-validator/challenges/challengeUtils.go  3
- packages/arb-validator/proofmachine/machinetester/machinetester.go  3
- packages/arb-validator/challenges/executionDefender.go  10
- packages/arb-validator/challenges/executionChallenger.go  10
- packages/arb-validator-core/ethbridgetest/value_test.go  6
         

Complexity decreasing per file
==============================
+ packages/arb-validator/loader/loader.go  -1
+ packages/arb-validator-core/ethbridge/arbRollup.go  -1
+ packages/arb-validator/challenges/testHelper.go  -10
+ packages/arb-validator/challenges/executionChallenge_test.go  -1
+ packages/arb-validator/challenges/inboxTopChallenge_test.go  -1
         

Clones added
============
- packages/arb-validator/challenges/inboxTopChallenger.go  3
- packages/arb-validator/challenges/messagesDefender.go  6
- packages/arb-validator-core/ethbridgetest/protocoltester/protocoltester.go  15
- packages/arb-validator/challenges/inboxTopDefender.go  2
- packages/arb-validator/proofmachine/proofmachine.go  1
- packages/arb-validator/challenges/messagesChallenger.go  6
- packages/arb-validator-core/ethbridgetest/messagetester/messagetester.go  2
- packages/arb-validator-core/ethbridgetest/valuetester/valuetester.go  5
- packages/arb-validator/challenges/challengeUtils.go  14
- packages/arb-validator/proofmachine/machinetester/machinetester.go  13
- packages/arb-validator/challenges/executionDefender.go  4
- packages/arb-validator/challenges/executionChallenger.go  4
- packages/arb-validator-core/ethbridgetest/value_test.go  2
- packages/arb-validator-core/ethbridge/messagesChallenge.go  2
         

Clones removed
==============
+ packages/arb-validator/challenges/testHelper.go  -1
+ packages/arb-validator/challenges/inboxTopChallenge_test.go  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5708250&bid=18682118)

File: packages/arb-validator/challenges/executionChallenger.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+	assertion, _ := mach.ExecuteAssertion(
+		totalSteps,
+		startPrecondition.TimeBounds,
+		startPrecondition.BeforeInbox,
+		0,
+	)
+	assertStub := valprotocol.NewExecutionAssertionStubFromAssertion(assertion)
+	precondition = precondition.GeneratePostcondition(assertStub)
+
+	return precondition
```
Comment (hkalodner) [inline packages/arb-validator/challenges/executionChallenger.go:166]: `return precondition.GeneratePostcondition(assertStub)`

File: packages/arb-bridge-eth/contracts/arch/Protocol.sol [hunk 2]
```diff
@@ -86,4 +68,26 @@ library Protocol {
             )
         );
     }
+
+    function generateLastMessageHash(
```
[the remark this answers is not in the document]
Comment (minhptruong) [reply packages/arb-bridge-eth/contracts/arch/Protocol.sol:72]: bad merge

File: packages/arb-validator/proofmachine/machinetester/machinetester.go [hunk 3]
```diff
@@ -0,0 +1,896 @@
+// Code generated - DO NOT EDIT.
```
[the remark this answers is not in the document]
Comment (minhptruong) [reply packages/arb-validator/proofmachine/machinetester/machinetester.go:1]: think the ethbridgetest is in arb-validator-core, which does not have a dependency on on the avm. Need that to test the machine

File: packages/arb-bridge-eth/contracts/vm/Staking.sol [hunk 4]
```diff
@@ -284,39 +292,74 @@ contract Staking is ChallengeType {
...
-        stakerCount--;
+    function _verifyStakerCount(
+        uint256[] memory stakerProofOffsets, 
+        uint256 _stakerCount
+    ) 
+        private 
+        view 
+    {
+        require(_stakerCount == stakerCount, CHCK_COUNT);
+        require(_stakerCount + 1 == stakerProofOffsets.length, CHCK_OFFSETS);
```
[the remark this answers is not in the document]
Comment (minhptruong) [reply packages/arb-bridge-eth/contracts/vm/Staking.sol:363]: true

**Pull request merged**